irohad/synchronizer/impl/synchronizer_impl.cpp

```diff
@@ -167,26 +164,55 @@ namespace iroha {
       }
     }
 
-    void SynchronizerImpl::processDifferent(const consensus::VoteOther &msg) {
-      log_->info("at handleDifferent");
-
-      shared_model::interface::types::HeightType top_block_height{0};
+    boost::optional<shared_model::interface::types::HeightType>
+    SynchronizerImpl::getTopBlockHeight() const {
+      decltype(getTopBlockHeight()) top_block_height;
       if (auto block_query = block_query_factory_->createBlockQuery()) {
         top_block_height = (*block_query)->getTopBlockHeight();
       } else {
         log_->error(
             "Unable to create block query and retrieve top block height");
+      }
+      return top_block_height;
+    }
+
+    void SynchronizerImpl::processDifferent(
+        const consensus::Synchronizable &msg,
+        bool process_small_height_difference,
+        shared_model::interface::types::HeightType expected_height,
+        SynchronizationOutcomeType alternative_outcome) {
+      log_->info("at handleDifferent");
+
+      auto top_block_height = getTopBlockHeight();
+      if (not top_block_height) {
+        log_->error("Unable to continue without knowing top block height");
         return;
       }
 
-      if (top_block_height >= msg.round.block_round) {
+      int64_t height_diff = msg.round.block_round - *top_block_height;
+      if (height_diff < 0) {
         log_->info(
             "Storage is already in synchronized state. Top block height is {}",
-            top_block_height);
+            *top_block_height);
+        return;
+      }
+
+      if (process_small_height_difference
+          and (0 == height_diff or 1 == height_diff)) {
+        notifier_.get_subscriber().on_next(
+            // TODO: nickaleks IR-147 18.01.19 add peers
+            // list from GateObject when it has one
+            SynchronizationEvent{
+                rxcpp::observable<>::empty<
+                    std::shared_ptr<shared_model::interface::Block>>(),
+                alternative_outcome,
+                msg.round,
+                nullptr /* this is going to be fixed in PR-2066 (IR-147) */});
         return;
       }
 
-      auto result = downloadMissingBlocks(msg, top_block_height);
+      auto result = downloadMissingBlocks(
+          msg, *top_block_height, expected_height, alternative_outcome);
```

Comment:
if we apply the suggestion above and this one, we can remove the `process_small_height_difference` and `expected_height` arguments of `processDifferent` and deduce them from `alternative_outcome`
```suggestion
          msg,
          *top_block_height,
          (alternative_outcome == SynchronizationOutcomeType::kCommit
               ? msg.round.block_round
               : msg.round.block_round - 1),
          alternative_outcome);
```